plugins/braintree-payment/src/providers/payment-braintree/src/core/braintree-base.ts

```diff
@@ -59,17 +62,114 @@ class BraintreeBase extends AbstractPaymentProvider<BraintreeOptions> {
   protected gateway: Braintree.BraintreeGateway;
   logger: Logger;
   container_: MedusaContainer;
+  clientToken: string | null;
+  cache: ICacheService;
   protected constructor(container: MedusaContainer, options: BraintreeOptions) {
     super(container, options);
 
     this.options_ = options;
     this.logger = container[ContainerRegistrationKeys.LOGGER];
     this.container_ = container;
-
+    this.cache = container[Modules.CACHE] as unknown as ICacheService;
     this.init();
   }
 
-  private parsePaymentSessionData(data: Record<string, unknown>): BraintreePaymentSessionData {
+  async saveClientTokenToCache(clientToken: string, customerId: string, expiresOn: number): Promise<void> {
+    if (!customerId) {
+      return;
+    }
+    if (expiresOn < 0) {
+      return;
+    }
+    if(expiresOn*1000 < Date.now()){
+      return;
+    }
+    if (!clientToken) {
+      return;
+    }
+    const expiryTime = expiresOn*1000 - Math.floor(Date.now())-1000;
+    if(expiryTime < 0){
+      return;
+    }
+    await this.cache.set(`braintree:clientToken:${customerId}`, clientToken, Math.floor(expiryTime / 1000));
+  }
+
+  async getClientTokenFromCache(customerId: string): Promise<string | null> {
+    const token = (await this.cache.get(`braintree:clientToken:${customerId}`)) as string | null;
+    if (token) {
+      this.clientToken = token;
+    }
+    return token;
+  }
+
+  async getValidClientToken(customerId: string): Promise<string | null> {
+    if (!customerId) {
+      const generatedToken = await this.gateway.clientToken.generate({});
+      return generatedToken.clientToken;
+    }
+    const token = await this.getClientTokenFromCache(customerId);
+    if (token) {
+      return token;
+    }
+    const generatedToken = await this.gateway.clientToken.generate({});
+    const expiresOn = this.getTokenExpiryTime(generatedToken);
+    await this.saveClientTokenToCache(generatedToken.clientToken, customerId, expiresOn);
+    return generatedToken.clientToken;
+  }
+  private decodeJWT(token: string): any {
+    try {
+      // Split the token into parts
+      const parts = token.split('.');
```

Comment:
@SGFGOV Man this seems like a lot of work. Why don't we just trust the braintree docs that say a token can be used up to 24 hours, instead of parsing this JWT? 

Instead of just storing the JWT as a string, we could store a small object in our cache:

`{token:string, created_at: string, expired_at: string}` and just store expires_at as 24 hours from now.

Then we would know when the token was created and when it expires, and not need to parse it for that.